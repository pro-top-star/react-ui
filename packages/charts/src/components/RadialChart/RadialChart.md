```jsx
import { RadialChart } from '@ui5/webcomponents-react/lib/RadialChart';
```
