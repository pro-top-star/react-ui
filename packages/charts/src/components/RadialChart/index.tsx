import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { Title } from '@ui5/webcomponents-react/lib/Title';
import { ChartOptions } from 'chart.js';
import React, { CSSProperties, FC, forwardRef, Ref, useMemo } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { ChartBaseDefaultProps } from '../../util/ChartBaseDefaultProps';
import { useMergedConfig } from '../../util/Utils';
import { DonutChart } from '../DonutChart';

export interface RadialChartPropTypes extends CommonProps {
  value: number;
  maxValue?: number;
  displayValue: number | string;
  colors?: Array<CSSProperties['color']>;
  options?: ChartOptions;
  height?: number;
  width?: number;
}

const styles = {
  radialChart: {
    position: 'relative'
  },
  content: {
    position: 'absolute',
    top: '0',
    left: '0',
    right: '0',
    bottom: '0',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    pointerEvents: 'none',
    fontSize: '2rem'
  }
};

const useStyles = createUseStyles(styles, { name: 'RadialChart' });

const RadialChart: FC<RadialChartPropTypes> = forwardRef((props: RadialChartPropTypes, ref: Ref<HTMLDivElement>) => {
  const { maxValue, value, displayValue, style, className, colors, options, width, height } = props;

  const data = [value, maxValue - value];
  const radialChartDefaultConfig = useMemo(() => {
    return {
      cutoutPercentage: 90,
      tooltips: {
        enabled: false
      },
      plugins: {
        datalabels: false
      }
    };
  }, []);
  const mergedOptions = useMergedConfig(radialChartDefaultConfig, options);

  const classes = useStyles();

  const radialChartContainerStyles = useMemo(
    () => ({
      width: `${width}px`,
      height: `${height}px`,
      ...style
    }),
    [width, height, style]
  );

  const outerClasses = StyleClassHelper.of(classes.radialChart);
  if (className) {
    outerClasses.put(className);
  }

  return (
    <div className={outerClasses.toString()} style={radialChartContainerStyles}>
      <DonutChart
        ref={ref}
        datasets={[{ data }]}
        colors={colors}
        options={mergedOptions}
        width={width}
        height={height}
        style={{ paddingTop: 0 }}
        noLegend
      />
      <div className={classes.content}>
        <Title>{displayValue as string}</Title>
      </div>
    </div>
  );
});

RadialChart.defaultProps = {
  ...ChartBaseDefaultProps,
  maxValue: 100,
  colors: ['#5899DA', '#adbcc3']
};

export { RadialChart };
