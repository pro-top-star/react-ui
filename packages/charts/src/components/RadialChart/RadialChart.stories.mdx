import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs/blocks';
import { DocsHeader } from '@shared/stories/DocsHeader';
import { RadialChart } from '@ui5/webcomponents-react-charts/dist/RadialChart';
import { DocsCommonProps } from '@shared/stories/DocsCommonProps';

<Meta
  title="Charts /  RadialChart"
  component={RadialChart}
  argTypes={{
    ...DocsCommonProps
  }}
  args={{
    displayValue: '50%',
    value: 50,
    maxValue: 100
  }}
/>

<DocsHeader />

## Example

<Canvas>
  <Story name="Default">{(props) => <RadialChart {...props} />}</Story>
</Canvas>

## Properties

<ArgsTable story="Default" />

<br />
<br />

## More Examples

### With Custom Color

<Canvas>
  <Story name="With Custom Color" args={{ color: '#f0ab00' }}>
    {(props) => <RadialChart {...props} />}
  </Story>
</Canvas>
