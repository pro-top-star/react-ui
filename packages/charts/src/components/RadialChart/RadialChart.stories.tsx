import { action } from '@storybook/addon-actions';
import { number, text } from '@storybook/addon-knobs';
import React from 'react';
import { RadialChart } from '../../lib/RadialChart';

export default {
  title: 'Charts /  RadialChart',
  component: RadialChart
};

export const renderStory = () => {
  return (
    <RadialChart
      value={number('value', 67)}
      displayValue={text('displayValue', '67%')}
      onDataPointClick={action('onDataPointClick')}
    />
  );
};

renderStory.story = {
  name: 'Default'
};

export const customColorStory = () => {
  return (
    <RadialChart
      value={number('value', 150)}
      displayValue={text('displayValue', '150l')}
      maxValue={number('maxValue', 250)}
      color={text('color', '#f0ab00')}
    />
  );
};

customColorStory.story = {
  name: 'with custom color'
};
