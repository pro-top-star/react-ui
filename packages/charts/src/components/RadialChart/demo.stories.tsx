import { number, text } from '@storybook/addon-knobs';
import { RadialChart } from '@ui5/webcomponents-react-charts/lib/RadialChart';
import React from 'react';
import notes from './RadialChart.md';

export default {
  title: 'Charts / RadialChart',
  component: RadialChart,
  parameters: { notes }
};

export const defaultStory = () => (
  <RadialChart value={number('value', 40)} displayValue={text('displayValue', '40%')} />
);

defaultStory.story = {
  name: 'Default'
};

export const inContainer = () => (
  <div style={{ width: '100px', height: '100px', margin: '0 auto', position: 'relative' }}>
    <RadialChart value={number('value', 90)} displayValue={text('displayValue', '90%')} width={100} height={100} />
  </div>
);

inContainer.story = {
  name: 'In Container'
};
