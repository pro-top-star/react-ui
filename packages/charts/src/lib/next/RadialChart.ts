import { RadialChart } from '../../components/RadialChart/RadialChart';

export { RadialChart };
