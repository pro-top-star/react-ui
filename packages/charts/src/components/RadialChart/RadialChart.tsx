import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import * as ThemingParameters from '@ui5/webcomponents-react-base/lib/sap_fiori_3';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { useInitialize } from '@ui5/webcomponents-react-charts/lib/initialize';
import { ChartContainer } from '@ui5/webcomponents-react-charts/lib/next/ChartContainer';
import { PieChartPlaceholder } from '@ui5/webcomponents-react-charts/lib/PieChartPlaceholder';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import React, { CSSProperties, FC, forwardRef, Ref, useCallback, useMemo } from 'react';
import { PolarAngleAxis, RadialBar, RadialBarChart as RadialBarChartLib } from 'recharts';

interface RadialChartProps extends CommonProps {
  value?: number;
  maxValue?: number;
  displayValue?: number | string;
  color?: CSSProperties['color'];
  onDataPointClick?: (event: Event) => void;
  height?: number | string;
  width?: number | string;
}

/**
 * <code>import { RadialChart } from '@ui5/webcomponents-react-charts/lib/next/RadialChart';</code>
 * **This component is under active development. The API is not stable yet and might change without further notice.**
 */
const RadialChart: FC<RadialChartProps> = forwardRef((props: RadialChartProps, ref: Ref<any>) => {
  const {
    maxValue = 100,
    value,
    displayValue,
    onDataPointClick,
    color,
    width = 300,
    height = 300,
    style,
    className,
    tooltip,
    slot
  } = props;

  useInitialize();

  const chartRef = useConsolidatedRef<any>(ref);

  const range = useMemo(() => {
    return [0, maxValue];
  }, [maxValue]);

  const dataset = useMemo(() => [{ value }], [value]);

  const onDataPointClickInternal = useCallback(
    (payload, i, event) => {
      if (payload && onDataPointClick) {
        onDataPointClick(
          Event.of(null, event, {
            value: payload.value,
            payload: payload.payload,
            xIndex: i
          })
        );
      }
    },
    [onDataPointClick]
  );

  return (
    <ChartContainer
      dataset={dataset}
      ref={chartRef}
      placeholder={PieChartPlaceholder}
      width={width as any}
      height={height}
      style={style}
      className={className}
      tooltip={tooltip}
      slot={slot}
    >
      <RadialBarChartLib
        margin={{ right: 30, top: 40, bottom: 30 }}
        innerRadius="90%"
        outerRadius="100%"
        barSize={10}
        data={dataset}
        cx="50%"
        cy="50%"
        startAngle={90}
        endAngle={-270}
      >
        <PolarAngleAxis type="number" domain={range} tick={false} />
        <RadialBar
          background={{ fill: ThemingParameters.sapContent_ImagePlaceholderBackground }}
          dataKey="value"
          cornerRadius="50%"
          fill={color ?? `var(--sapUiChartAccent${(0 % 12) + 1})`}
          onClick={onDataPointClickInternal}
        />
        <text
          x="50%"
          y="50%"
          textAnchor="middle"
          dominantBaseline="middle"
          className="progress-label"
          style={{ fontSize: ThemingParameters.sapMFontHeader3Size, fill: ThemingParameters.sapTextColor }}
        >
          {displayValue}
        </text>
      </RadialBarChartLib>
    </ChartContainer>
  );
});

RadialChart.displayName = 'RadialChart';

export { RadialChart };
