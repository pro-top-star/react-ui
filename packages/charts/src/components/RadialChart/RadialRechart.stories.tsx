import { action } from '@storybook/addon-actions';
import { number, text } from '@storybook/addon-knobs';
import { RadialChart } from '@ui5/webcomponents-react-charts/lib/next/RadialChart';
import React from 'react';

const dataset = [{ name: 'L1', value: 67 }];

export default {
  title: 'Charts - Unstable /  RadialChart',
  component: RadialChart
};

export const renderStory = () => {
  return (
    <RadialChart
      value={number('value', 67)}
      displayValue={text('displayValue', '67%')}
      width={text('width', '300px')}
      height={text('height', '300px')}
      onDataPointClick={action('onDataPointClick')}
    />
  );
};

renderStory.story = {
  name: 'Default'
};

export const customColorStory = () => {
  return (
    <RadialChart
      value={number('value', 150)}
      displayValue={text('displayValue', '150l')}
      maxValue={number('maxValue', 250)}
      width={text('width', '300px')}
      height={text('height', '300px')}
      color={text('color', '#f0ab00')}
    />
  );
};

customColorStory.story = {
  name: 'with custom color'
};
