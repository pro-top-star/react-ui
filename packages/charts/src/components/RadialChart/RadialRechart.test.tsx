import { mountThemedComponent, renderThemedComponent } from '@shared/tests/utils';
import * as React from 'react';
import { RadialChart } from './RadialChart';

describe('RadialChart', () => {
  test('Renders with data', () => {
    renderThemedComponent(<RadialChart width={'100%'} value={67} displayValue={'67%'} />);
  });

  test('loading placeholder', () => {
    const wrapper = mountThemedComponent(<RadialChart width={'50%'} />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
