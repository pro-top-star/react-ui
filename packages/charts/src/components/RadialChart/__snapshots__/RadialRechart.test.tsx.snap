// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadialChart loading placeholder 1`] = `
<div
  style="width: 50%; height: 300px;"
>
  <div
    class="recharts-responsive-container"
    id="undefined"
    style="width: 100%; height: 100%;"
  >
    <div />
  </div>
</div>
`;
